api/utxoCache.ts

```diff
@@ -0,0 +1,134 @@
+      return JSON.parse(cache);
+    } else {
+      return {
+        version: -1,
+        utxos: {},
+      };
+    }
+  };
+
+  private _setCache = async (utxos: UtxoCacheStruct, address: string): Promise<void> => {
+    await this._cacheStorageController.set(
+      this.getAddressCacheStorageKey(address),
+      JSON.stringify({
+        version: this.VERSION,
+        utxos: utxos,
+      }),
+    );
+  };
+
+  _getAddressUtxos = async (address: string) => getAddressUtxoOrdinalBundles(address, 0, 60);
+
+  _getUtxo = async (txid: string, vout: number) => getUtxoOrdinalBundle(txid, vout);
+
+  clear = async (address: string): Promise<void> => {
+    await this._cacheStorageController.remove(this.getAddressCacheStorageKey(address));
+  };
+
+  getVersion = async (address: string): Promise<number> => {
+    const { version } = await this._getCache(address);
+    if (version) {
+      return version;
+    } else {
+      return -1;
+    }
+  };
+
+  isCacheUpdated = async (address: string): Promise<boolean> => {
+    const version = await this.getVersion(address);
+    return version === this.VERSION;
+  };
+
+  getUtxoState = async (utxoId: string, address: string): Promise<UtxoOrdinalBundle | null> => {
+    const { utxos } = await this._getCache(address);
+    console.log('🚀 ~ file: utxoCache.ts:82 ~ UtxoCache ~ getUtxoState= ~ utxos:', utxos);
+    if (utxos && utxos[utxoId]) {
+      return utxos[utxoId];
+    }
+    const [txid, vout] = utxoId.split(':');
+    const utxo = await this._getUtxo(txid, parseInt(vout, 10));
+    await this.setUtxoState(utxoId, utxo, address, utxos);
+    return utxo;
+  };
+
+  setUtxoState = async (
+    utxoId: string,
+    utxo: UtxoOrdinalBundle,
+    address: string,
+    cachedUtxos: UtxoCacheStruct,
+  ): Promise<void> => {
+    console.log('🚀 ~ file: utxoCache.ts:97 ~ UtxoCache ~ cachedUtxos:', cachedUtxos);
+    console.log('🚀 ~ file: utxoCache.ts:97 ~ UtxoCache ~ address:', address);
+    console.log('🚀 ~ file: utxoCache.ts:97 ~ UtxoCache ~ utxo:', utxo);
+    console.log('🚀 ~ file: utxoCache.ts:97 ~ UtxoCache ~ utxoId:', utxoId);
+    if (!cachedUtxos) {
+      cachedUtxos = {};
+    }
+    cachedUtxos[utxoId] = utxo;
+    await this._setCache(this.utxos, address);
+  };
+
+  removeUtxoState = async (utxoId: string, address: string): Promise<void | boolean> => {
+    const { utxos: cachedUtxos } = await this._getCache(address);
+    if (cachedUtxos && utxoId in cachedUtxos) {
+      delete cachedUtxos[utxoId];
+      await this._setCache(cachedUtxos, address);
+    }
+    return false;
+  };
+
+  getAllUtxos = async (btcAddress: string): Promise<UtxoCacheStruct> => {
+    const utxos = await this._getAddressUtxos(btcAddress);
+    const utxosObject: UtxoCacheStruct = utxos.results.reduce((acc, utxo) => {
+      acc[`${utxo.txid}:${utxo.vout}`] = utxo;
+      return acc;
+    }, {} as UtxoCacheStruct);
+    return utxosObject;
+  };
+
+  initCache = async (address: string): Promise<void> => {
```

Comment:
I think there are too many functions exposed from this cache. The interface should be as simple as possible, so the only function exposed should be `getUtxoState` which returns the state for a single UTXO. The user of this class shouldn't need to manually check versions, initialise, and clear the cache.

That function should:
- try get the cache for the address
- if it doesn't exist or is wrong version, it re-initialise the cache, store it, and return the required UTXO
- if it does exist, load it and check if the UTXO is in the cache for the address
- if it is, return the state, if not, get that specific UTXO from the API, store it in the cache, and return the state

The getAllUtxos is a bit dangerous because the cache could be out of date and we might not have the latest UTXOs cached. Best to just allow a query to a single UTXO at a time.

Reply:
It might actually be nice if we could initialise the cache in an in-memory mode with a wallet (so either 1 or 2 addresses) which it loads into memory. When we call the getUtxo method, it just works with the stuff in-memory and only stores those lookup objects into the cache when there are changes. If it's called directly with an address that's not in memory, it tries to load from storage for that address.

Alternatively, it could just keep the state in-memory for a max of 2 addresses and not need initialisation.